### PR TITLE
Feat: Kl Divergence kernel

### LINKED ISSUE #188: Support KL divergence kernel
### 🚀 The feature, motivation and pitch

Often used in RLHF. As requested by axolotl team

https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by S1ro1**
## Summary
Implements KL Divergence loss function custom kernel. Resolves #188 .

## Testing Done
Implemented tests that cross-check the results of the forward/backward pass compared to `torch.nn.KLDivLoss`. Also added script to benchmark the memory and speed performance of the new Liger implementation in comparison to Torch native one, with results as shown below.

image

image

- Hardware Type: Nvidia RTX A5500/Nvidia H100 (80GB PCIe)
- [x] run `make test` to ensure correctness
- [x] run `make checkstyle` to ensure code style
- [x] run `make test-convergence` to ensure convergence

### CODE REVIEW

**Comment by S1ro1** [comment]
The implementation isn't perfect I think, wanna study a bit more how to optimize it, but since it outperforms torch across all the vocab sizes, I feel like it can be merged as is for now.

**Comment by S1ro1** [comment]
Also a note, so many tests are added to the runners because of many combinations (different reductions/log_target) which all change the calculation done. I tried to keep the sizes on the smaller side to not waste too much test time for that reason.

**Comment by ByronHsu** [comment]
cc @lancerts to validate the correctness and code style

File: src/liger_kernel/ops/kl_div.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            res = -tl.exp(target)
+
+        tl.store(input_ptr + offsets, res, mask=mask)
+
+
+def kldiv_forward_triton(y_pred, y_true, log_target, reduction):  # [B, S]  # [B, S]
+    B, S = y_pred.shape
+
+    BLOCK_SIZE = min(MAX_FUSED_SIZE, triton.next_power_of_2(S))
+    num_warps = get_num_warps(BLOCK_SIZE)
```
Comment (yundai424) [inline src/liger_kernel/ops/kl_div.py:116]: perhaps can reuse https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/ops/utils.py#L23

Comment (S1ro1) [reply src/liger_kernel/ops/kl_div.py:116]: Actually cannot, since the shape is BT*V, max block size would be exceeded pretty often, for that reason the custom logic was used

File: test/transformers/test_kl_div.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    x2 = input.detach().clone().requires_grad_(True)
+
+    with torch.no_grad():
+        target = torch.randn(B * T, V, device="cuda").softmax(dim=-1)
+
+    output = torch_kldiv(x1, target)
+    output2 = target_kldiv(x2, target)
+    assert_verbose_allclose(output, output2, atol=atol, rtol=rtol)
+
+    if not is_last:
```
Comment (qingquansong) [inline test/transformers/test_kl_div.py:64]: Out of curiosity, what is the `is_last` here used for?

Comment (S1ro1) [reply test/transformers/test_kl_div.py:64]: If not is_last, I multiply the result to do more calculations for grad, since if it's the last operation we can skip one multiply in backwards calculation, since the output grads are 1.0

File: src/liger_kernel/ops/kl_div.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    base_offsets = tl.arange(0, BLOCK_SIZE)
+
+    for i in range(0, n_cols, BLOCK_SIZE):
+        offsets = i + base_offsets
+        mask = offsets < n_cols
+        y = tl.load(y_ptr + offsets, mask=mask, other=0.0)
+        y_true = tl.load(gt_ptr + offsets, mask=mask, other=0.0)
+
+        if not log_target:
+            loss = y_true * (tl.log(y_true) - y)
```
Comment (lancerts) [inline src/liger_kernel/ops/kl_div.py:67]: can we add the math formula in the comment?
KL(p, q) = sum (p_i * log(pi/qi)), are we computing KL(y, y_true) or KL(y_true, y)?
is y the probability of the logits?

Lets add details in the comments, thanks

File: test/transformers/test_kl_div.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+
+def _test_correctness_once(
+    target_kldiv, B, T, V, dtype, atol, rtol, reduction, log_target, is_last=True
+):
+    torch.manual_seed(0)
+    torch_kldiv = KLDivLoss(reduction=reduction, log_target=log_target)
+
+    input = torch.randn(
+        B * T, V, device="cuda", dtype=dtype, requires_grad=True
```
Comment (lancerts) [inline test/transformers/test_kl_div.py:51]: make device as an arg instead of "cuda" since people may want to use it on AMD GPU

File: src/liger_kernel/ops/kl_div.py (hunk 3 above)
Comment (lancerts) [reply src/liger_kernel/ops/kl_div.py:67]: Its good also to include "https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html" as the reference since `if not log_target` is the same as the pytorch one. If we are using a different reference, feel free to link that one in the comments, thanks

File: src/liger_kernel/ops/kl_div.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        output_tensor.stride(0),
+        S,
+        BLOCK_SIZE=BLOCK_SIZE,
+        num_warps=num_warps,
+        log_target=log_target,
+        reduction=reduction,
+    )
+
+    if reduction == _REDUCTION_MODE_BATCHMEAN:
+        return output_tensor.sum() / B
```
Comment (lancerts) [inline src/liger_kernel/ops/kl_div.py:139]: Is  BATCHMEAN output_tensor.sum() / B and MEAN output_tensor.sum() / (BXT) in our case? lets add a comment and explain whether it is aligned with https://pytorch.org/docs/stable/generated/torch.nn.KLDivLoss.html, ty

Comment (lancerts) [reply src/liger_kernel/ops/kl_div.py:139]: https://github.com/pytorch/pytorch/blob/d7b57c4d63edb42e1deeeba9497fcb5f1f748ff2/torch/nn/functional.py#L3372

**Review by S1ro1 — COMMENTED** [review]
> https://github.com/pytorch/pytorch/blob/d7b57c4d63edb42e1deeeba9497fcb5f1f748ff2/torch/nn/functional.py#L3372 

Yeah, requested info on that on discord before, but probably dissapeared between conversations, should I change it to reflect now?

**Comment by lancerts** [comment]
> > https://github.com/pytorch/pytorch/blob/d7b57c4d63edb42e1deeeba9497fcb5f1f748ff2/torch/nn/functional.py#L3372
> 
> Yeah, requested info on that on discord before, but probably dissapeared between conversations, should I change it to reflect now?

Code is fine, lets add explicitly the link to pytorch as our implementation is "inconsistent with torch" and will change accordingly with torch in the future.

**Comment by S1ro1** [comment]
@lancerts I think I have addressed all of the comments, should be good to go now.

File: src/liger_kernel/ops/kl_div.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+    base_offsets = tl.arange(0, BLOCK_SIZE)
+
+    for i in range(0, n_cols, BLOCK_SIZE):
+        offsets = i + base_offsets
+        mask = offsets < n_cols
+        y = tl.load(y_ptr + offsets, mask=mask, other=0.0)
+        y_true = tl.load(gt_ptr + offsets, mask=mask, other=0.0)
+
+        # KL(P || Q) = P * (log(P) - Q), however, the arguments to function are in reverse order, to match the Pytorch Loss function API
```
Comment (lancerts) [inline src/liger_kernel/ops/kl_div.py:66]: KL(P || Q) = P * (log(P) - logQ), and we are doing KL(y_true || y )

File: src/liger_kernel/ops/kl_div.py (hunk 5 above)
Comment (lancerts) [reply src/liger_kernel/ops/kl_div.py:143]: add https://github.com/pytorch/pytorch/blob/d7b57c4d63edb42e1deeeba9497fcb5f1f748ff2/torch/nn/functional.py#L3372 link also

File: test/transformers/test_kl_div.py (hunk 2 above)
Comment (lancerts) [reply test/transformers/test_kl_div.py:74]: add a comment to explain: is_last = is_last_layer?

File: src/liger_kernel/ops/kl_div.py (hunk 1 above)
Comment (yundai424) [reply src/liger_kernel/ops/kl_div.py:118]: Ah yup true!

File: src/liger_kernel/ops/kl_div.py (hunk 6 above)
Comment (S1ro1) [reply src/liger_kernel/ops/kl_div.py:66]: Correct, fixed now.

Comment (lancerts) [reply src/liger_kernel/ops/kl_div.py:66]: KL(P || Q) = P * (log(P) - Q), i guess it is log Q?

File: src/liger_kernel/ops/kl_div.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+    "none": _REDUCTION_MODE_NONE.value,
+    "sum": _REDUCTION_MODE_SUM.value,
+    "mean": _REDUCTION_MODE_MEAN.value,
+    "batchmean": _REDUCTION_MODE_BATCHMEAN.value,
+}
+
+
+@triton.jit
+def _kldiv_kernel_forward(
+    y_ptr,  # [B, S], prediction ptr
```
Comment (lancerts) [inline src/liger_kernel/ops/kl_div.py:41]: please add another comment "input y should be a distribution in the log space"

**Comment by lancerts** [comment]
Just a few comments, other lgtm

**Review by lancerts — APPROVED** [review]

**Comment by ByronHsu** [comment]
Please add kl div to the readme too
